### PR TITLE
Better handle postgres requests for wallet objects

### PR DESCRIPTION
**Pull request by NicolasDorier**
Because of the way we handle upsert through EF, the logs in postgres seems scary:

```
2023-05-17 09:28:19.834 UTC [9293] ERROR:  duplicate key value violates unique constraint "PK_WalletObjects"
2023-05-17 09:28:19.834 UTC [9293] DETAIL:  Key ("WalletId", "Type", "Id")=(S-AprSc64pgef2z67QkY9SecxB2...-BTC, label, invoice) already exists.
2023-05-17 09:28:19.834 UTC [9293] STATEMENT:  INSERT INTO "WalletObjects" ("Id", "Type", "WalletId", "Data")
        VALUES ($1, $2, $3, $4)
2023-05-17 09:28:19.844 UTC [9298] ERROR:  duplicate key value violates unique constraint "PK_WalletObjects"
2023-05-17 09:28:19.844 UTC [9298] DETAIL:  Key ("WalletId", "Type", "Id")=(S-AprSc64pgef2z67QkY9SecxB2...-BTC, invoice, KuezCbYieRQPJaP2irr6fm) already exists.
2023-05-17 09:28:19.844 UTC [9298] STATEMENT:  INSERT INTO "WalletObjects" ("Id", "Type", "WalletId", "Data")
        VALUES ($1, $2, $3, $4)
2023-05-17 09:28:19.855 UTC [9294] ERROR:  duplicate key value violates unique constraint "PK_WalletObjects"
2023-05-17 09:28:19.855 UTC [9294] DETAIL:  Key ("WalletId", "Type", "Id")=(S-AprSc64pgef2z67QkY9SecxB2...-BTC, tx, 05aaac4b...) already exists.
2023-05-17 09:28:19.855 UTC [9294] STATEMENT:  INSERT INTO "WalletObjects" ("Id", "Type", "WalletId", "Data")
        VALUES ($1, $2, $3, $4)
2023-05-17 09:32:51.961 UTC [9327] ERROR:  duplicate key value violates unique constraint "PK_WalletObjects"
2023-05-17 09:32:51.961 UTC [9327] DETAIL:  Key ("WalletId", "Type", "Id")=(S-AprSc64pgef2z67QkY9SecxB2...-BTC, tx, b05de59...) already exists.
2023-05-17 09:32:51.961 UTC [9327] STATEMENT:  INSERT INTO "WalletObjects" ("Id", "Type", "WalletId", "Data")
        VALUES ($1, $2, $3, $4)
2023-05-17 09:32:51.965 UTC [9329] ERROR:  duplicate key value violates unique constraint "PK_WalletObjectLinks"
2023-05-17 09:32:51.965 UTC [9329] DETAIL:  Key ("WalletId", "AType", "AId", "BType", "BId")=(S-AprSc64pgef2z67QkY9SecxB2...-BTC, address, 38vfg8sJhwPDCxpobHvRH6GCCLGgGPhYYs, tx, b05de59...) already exists.
2023-05-17 09:32:51.965 UTC [9329] STATEMENT:  INSERT INTO "WalletObjectLinks" ("AId", "AType", "BId", "BType", "WalletId", "Data")
        VALUES ($1, $2, $3, $4, $5, $6)
2023-05-17 09:32:51.972 UTC [9330] ERROR:  duplicate key value violates unique constraint "PK_WalletObjects"
2023-05-17 09:32:51.972 UTC [9330] DETAIL:  Key ("WalletId", "Type", "Id")=(S-AprSc64pgef2z67QkY9SecxB2...-BTC, tx, b8e4f81c9d3effe1744249f782cc01d6cde8ea12f5d749c1b97e39aa3196e9a5) already exists.
2023-05-17 09:32:51.972 UTC [9330] STATEMENT:  INSERT INTO "WalletObjects" ("Id", "Type", "WalletId", "Data")
        VALUES ($1, $2, $3, $4)
2023-05-17 09:32:51.975 UTC [9357] ERROR:  duplicate key value violates unique constraint "PK_WalletObjectLinks"
2023-05-17 09:32:51.975 UTC [9357] DETAIL:  Key ("WalletId", "AType", "AId", "BType", "BId")=(S-AprSc64pgef2z67QkY9SecxB2...-BTC, address, 3GLwGRJsVyh6msebSmC6t86Q45u9gYeyPM, tx, b8e4f8...) already exists.
2023-05-17 09:32:51.975 UTC [9357] STATEMENT:  INSERT INTO "WalletObjectLinks" ("AId", "AType", "BId", "BType", "WalletId", "Data")
        VALUES ($1, $2, $3, $4, $5, $6)
2023-05-17 09:32:51.982 UTC [9327] ERROR:  duplicate key value violates unique constraint "PK_WalletObjects"
2023-05-17 09:32:51.982 UTC [9327] DETAIL:  Key ("WalletId", "Type", "Id")=(S-AprSc64pgef2z67QkY9SecxB2...-BTC, tx, 05aaac...) already exists.
2023-05-17 09:32:51.982 UTC [9327] STATEMENT:  INSERT INTO "WalletObjects" ("Id", "Type", "WalletId", "Data")
        VALUES ($1, $2, $3, $4)
2023-05-17 09:32:51.985 UTC [9329] ERROR:  duplicate key value violates unique constraint "PK_WalletObjectLinks"
2023-05-17 09:32:51.985 UTC [9329] DETAIL:  Key ("WalletId", "AType", "AId", "BType", "BId")=(S-AprSc64pgef2z67QkY9SecxB2...-BTC, address, 3DyFYN2kru7oTWLCivAFHpCL9ipjV2bhei, tx, 05aaac...) already exists.
2023-05-17 09:32:51.985 UTC [9329] STATEMENT:  INSERT INTO "WalletObjectLinks" ("AId", "AType", "BId", "BType", "WalletId", "Data")
        VALUES ($1, $2, $3, $4, $5, $6)
2023-05-17 09:36:18.534 UTC [9328] ERROR:  duplicate key value violates unique constraint "PK_WalletObjects"
2023-05-17 09:36:18.534 UTC [9328] DETAIL:  Key ("WalletId", "Type", "Id")=(S-AprSc64pgef2z67QkY9SecxB2...-BTC, label, invoice) already exists.
2023-05-17 09:36:18.534 UTC [9328] STATEMENT:  INSERT INTO "WalletObjects" ("Id", "Type", "WalletId", "Data")
        VALUES ($1, $2, $3, $4)
2023-05-17 09:36:18.543 UTC [9329] ERROR:  duplicate key value violates unique constraint "PK_WalletObjects"
2023-05-17 09:36:18.543 UTC [9329] DETAIL:  Key ("WalletId", "Type", "Id")=(S-AprSc64pgef2z67QkY9SecxB2...-BTC, invoice, 7td1y7BcmUSz12WcLUcjB3) already exists.
2023-05-17 09:36:18.543 UTC [9329] STATEMENT:  INSERT INTO "WalletObjects" ("Id", "Type", "WalletId", "Data")
        VALUES ($1, $2, $3, $4)
2023-05-17 09:36:18.550 UTC [9313] ERROR:  duplicate key value violates unique constraint "PK_WalletObjects"
2023-05-17 09:36:18.550 UTC [9313] DETAIL:  Key ("WalletId", "Type", "Id")=(S-AprSc64pgef2z67QkY9SecxB2...-BTC, tx, 3da0abc...) already exists.
2023-05-17 09:36:18.550 UTC [9313] STATEMENT:  INSERT INTO "WalletObjects" ("Id", "Type", "WalletId", "Data")
        VALUES ($1, $2, $3, $4)`
```

This isn't a problem per say, but this PR fix it by doing a proper postgres native upsert rather than hacking around entity framework.

Fix https://github.com/btcpayserver/btcpayserver/issues/4977

### CODE REVIEW

File: BTCPayServer/Services/WalletRepository.cs [hunk 1]
```diff
@@ -374,13 +380,27 @@ public async Task EnsureWalletObjectLink(WalletObjectId a, WalletObjectId b, JOb
...
+                        e.State = EntityState.Modified;
+                        await ctx.SaveChangesAsync();
+                    }
+                }
             }
-            catch (DbUpdateException) // already exists
+            else
             {
+                var connection = ctx.Database.GetDbConnection();
+                var conflict = doNothingIfExists ? "ON CONFLICT DO NOTHING" : "ON CONFLICT DO UPDATE SET \"Data\"=EXCLUDED.\"Data\"";
```
Comment (Kukks) [inline BTCPayServer/Services/WalletRepository.cs:402]: rofl at this Postgres syntax. They really went verbose huh?

**Review by Kukks — COMMENTED** [review]
I'm really scared of all these raw SQL constructions scattered everywhere. It seems like it is hard to  track them down when doing a schema change, and things will start breaking.

File: BTCPayServer/Services/WalletRepository.cs (hunk 1 above)
Comment (NicolasDorier) [reply BTCPayServer/Services/WalletRepository.cs:402]: it's not really postgres fault, this is SQL standard.

**Comment by NicolasDorier** [comment]
> I'm really scared of all these raw SQL constructions scattered everywhere. It seems like it is hard to track them down when doing a schema change, and things will start breaking.

That's why we have tests :p

**Comment by NicolasDorier** [comment]
Tests pass now. I think next big release we should cleanup those postgres specific if/else, since we don't support mysql and sqlite anymore.

**Review by Kukks — APPROVED** [review]

**Pull request merged**